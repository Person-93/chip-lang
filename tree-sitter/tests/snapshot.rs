use chip_snapshot_tests::{assert_snapshot, run_cases};
use std::fmt::{self, Write};
use tree_sitter::{Parser, Point, Tree, TreeCursor};

fn main() {
  let mut parser = Parser::new();
  parser
    .set_language(tree_sitter_chip::language())
    .expect("assign language to parser");

  run_cases(|source| {
    parser.reset();
    let tree = parser
      .parse(&source, None)
      .expect("parse snapshot test case");
    assert_snapshot!(print(&tree, &source).trim());
  });
}

pub fn print(tree: &Tree, source: &str) -> String {
  let mut output = String::new();
  let mut cursor = tree.walk();
  output
    .print_node(&mut cursor, source, 0)
    .expect("debug format tree");
  output
}

impl<T: Write> NodePrinter for T {}

trait NodePrinter: Write {
  fn print_node(
    &mut self,
    cursor: &mut TreeCursor,
    source: &str,
    indentation: u8,
  ) -> fmt::Result {
    let node = cursor.node();
    let kind = node.kind();
    let field_name = cursor.field_name();
    let Point { row, column } = node.start_position();
    let Point {
      row: end_row,
      column: end_column,
    } = node.end_position();

    for _ in 0..indentation {
      write!(self, "  ")?;
    }
    if node.is_missing() {
      write!(self, "MISSING ")?;
    }
    if let Some(field_name) = field_name {
      write!(self, "{field_name}:")?;
    }
    write!(self, "{kind}")?;
    write!(self, " @ {row}:{column}..{end_row}:{end_column}")?;

    if (node.is_named() || field_name.is_some()) && node.child_count() == 0 {
      let value = node.utf8_text(source.as_bytes()).unwrap();
      if kind != value {
        write!(self, "\t{}", node.utf8_text(source.as_bytes()).unwrap())?;
      }
    }

    writeln!(self)?;

    if cursor.goto_first_child() {
      self.print_node(&mut cursor.clone(), source, indentation + 1)?;
      while cursor.goto_next_sibling() {
        self.print_node(&mut cursor.clone(), source, indentation + 1)?;
      }
    }

    Ok(())
  }
}
