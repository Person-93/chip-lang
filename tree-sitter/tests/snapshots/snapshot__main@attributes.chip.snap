---
source: tree-sitter/tests/snapshot.rs
assertion_line: 16
expression: "print(&tree, &source).trim()"
input_file: tree-sitter/bindings/rust/testing/cases/attributes.chip
---
file @ 0:0..11:0
  #^[ @ 0:0..0:3
  attr @ 0:3..0:9
    path @ 0:3..0:9
      first_segment:identifier @ 0:3..0:9	simple
  ] @ 0:9..0:10
  #^[ @ 1:0..1:3
  attr @ 1:3..1:9
    path @ 1:3..1:9
      rooted::: @ 1:3..1:5
      first_segment:identifier @ 1:5..1:9	attr
  ] @ 1:9..1:10
  #^[ @ 2:0..2:3
  attr @ 2:3..2:22
    path @ 2:3..2:22
      first_segment:identifier @ 2:3..2:7	attr
      :: @ 2:7..2:9
      segment:identifier @ 2:9..2:13	with
      :: @ 2:13..2:15
      segment:identifier @ 2:15..2:16	a
      :: @ 2:16..2:18
      segment:identifier @ 2:18..2:22	path
  ] @ 2:22..2:23
  #^[ @ 3:0..3:3
  attr @ 3:3..3:19
    path @ 3:3..3:9
      first_segment:identifier @ 3:3..3:9	string
    = @ 3:10..3:11
    value:string_literal @ 3:12..3:19	"hello"
  ] @ 3:19..3:20
  #^[ @ 4:0..4:3
  attr @ 4:3..4:9
    path @ 4:3..4:4
      first_segment:identifier @ 4:3..4:4	n
    = @ 4:5..4:6
    value:numeric_literal @ 4:7..4:9
      digits:digits @ 4:7..4:9	42
  ] @ 4:9..4:10
  #^[ @ 5:0..5:3
  attr @ 5:3..5:11
    path @ 5:3..5:4
      first_segment:identifier @ 5:3..5:4	b
    = @ 5:5..5:6
    value:boolean_literal @ 5:7..5:11
      true @ 5:7..5:11
  ] @ 5:11..5:12
  #^[ @ 6:0..6:3
  attr @ 6:3..6:32
    path @ 6:3..6:7
      first_segment:identifier @ 6:3..6:7	call
    ( @ 6:7..6:8
    arg:string_literal @ 6:8..6:14	"arg0"
    , @ 6:14..6:15
    arg:attr @ 6:16..6:20
      path @ 6:16..6:20
        first_segment:identifier @ 6:16..6:20	arg1
    , @ 6:20..6:21
    arg:numeric_literal @ 6:22..6:24
      digits:digits @ 6:22..6:24	42
    , @ 6:24..6:25
    arg:boolean_literal @ 6:26..6:31
      false @ 6:26..6:31
    ) @ 6:31..6:32
  ] @ 6:32..6:33
  #^[ @ 7:0..7:3
  attr @ 7:3..7:43
    path @ 7:3..7:8
      first_segment:identifier @ 7:3..7:8	call2
    ( @ 7:8..7:9
    arg:attr @ 7:9..7:13
      path @ 7:9..7:13
        first_segment:identifier @ 7:9..7:13	arg0
    , @ 7:13..7:14
    arg:attr @ 7:15..7:27
      path @ 7:15..7:21
        first_segment:identifier @ 7:15..7:21	nested
      ( @ 7:21..7:22
      arg:attr @ 7:22..7:26
        path @ 7:22..7:26
          first_segment:identifier @ 7:22..7:26	call
      ) @ 7:26..7:27
    , @ 7:27..7:28
    arg:attr @ 7:29..7:42
      path @ 7:29..7:33
        first_segment:identifier @ 7:29..7:33	this
      = @ 7:34..7:35
      value:string_literal @ 7:36..7:42	"that"
    ) @ 7:42..7:43
  ] @ 7:43..7:44
  function @ 9:0..10:22
    #[ @ 9:0..9:2
    attr @ 9:2..9:11
      path @ 9:2..9:11
        first_segment:identifier @ 9:2..9:11	some_attr
    ] @ 9:11..9:12
    fn @ 10:0..10:2
    name:identifier @ 10:3..10:16	fn_with_attrs
    ( @ 10:16..10:17
    ) @ 10:17..10:18
    : @ 10:18..10:19
    body:codeblock @ 10:20..10:22
      { @ 10:20..10:21
      } @ 10:21..10:22
